### PR TITLE
Add timeout to warm gittar cache step

### LINKED ISSUE #314: Time out when gittar cache warming hangs
It seems that github can be very unresponsive sometimes, which will cause installation (upgrade) of the zkapp CLI or its examples to hang (in my experience, sometimes indefinitely) during the step `Warm cache for project template.`. Since warming the cache is not a must have, we should have a timeout on this operation, so the user can proceed with installation when it doesn't work.

### PR DESCRIPTION
**Pull request by ymekuria**
Closes #314 

**Background**

Sometimes the installation or upgrade of the zkapp-cli hangs during the `warm cache for project template` step in the `postinstall.js` file when a network request is made by gittar to fetch the zkapp-cli templates and examples from github [here](https://github.com/o1-labs/zkapp-cli/blob/main/src/postinstall.js#L59).

**Approach**

This PR adds a timeout to the fetch gittar operation. If the fetch operation does not resolve before the the timeout, the `warm cache for project template` step is skipped, and the installation continues. See [this](https://github.com/o1-labs/zkapp-cli/issues/314) discussion for more details. 

**Next steps**

A log was added to show which NPM packages were added to the cache in the `warm NPM cache for project template deps` step. A log was also added before the gittar fetch operation in the `warm cache for project template` step. We can use these logs to gain better visibility into hanging operations in the future and adjust the approach to addressing them accordingly.

### CODE REVIEW

File: src/postinstall.js [hunk 1]
```diff
@@ -55,9 +55,35 @@ async function warmGittarCache() {
   console.log('  Warm cache for project template.');
 
   try {
+    const maxTimeLimit = 7000;
```
Comment (ymekuria) [inline src/postinstall.js:58]: This timeout limit is debatable.

**Comment by jasongitmail** [comment]
Important for an eng reviewer to actually pull this branch to test it please

File: src/postinstall.js (hunk 1 above)
Comment (jasongitmail) [reply src/postinstall.js:58]: We should set this based on how long it takes to succeed for users with poor internet. So I'd go much higher personally--maybe 15sec

**Review by nicc — APPROVED** [review]
Looks good. Tested locally by dropping the timeout to 1ms and installing from my branch. 
fyi to see the relevant logs I had to npm install with `--verbose --foreground-scripts`

**Comment by ymekuria** [comment]
>fyi to see the relevant logs I had to npm install with `--verbose --foreground-scripts`

Interesting. Were you able to see other logs from the `postinstall.js` file without the --verbose `--foreground-scripts` flag?

File: src/postinstall.js [hunk 2]
```diff
@@ -42,8 +41,20 @@ async function warmNpmCache() {
...
+    const addPkgMaxTimeLimit = 1500;
     for await (const pkgWithVersion of toCache) {
-      await shExec(`npm cache add ${pkgWithVersion}`);
+      console.log(`  Adding ${pkgWithVersion} to the cache.`);
+      // Skips "Warm NPM cache for project template deps" steps  if adding
+      // a pacakage takes longer than the add package max time limit.
+      const npmCacheResponse = await executeInTimeLimit(
+        shExec(`npm cache add ${pkgWithVersion}`),
+        addPkgMaxTimeLimit
+      );
```
Comment (ymekuria) [inline src/postinstall.js:52]: I made the time function more general and added a timeout of 1500 ms to add a package to the `npm cache`.

**Pull request merged**